lib/trino-hdfs/src/main/java/io/trino/hdfs/gcs/GcsExclusiveOutputStream.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.hdfs.gcs;
+
+import com.google.cloud.hadoop.repackaged.gcs.com.google.api.client.http.ByteArrayContent;
+import com.google.cloud.hadoop.repackaged.gcs.com.google.api.services.storage.Storage;
+import com.google.cloud.hadoop.repackaged.gcs.com.google.api.services.storage.model.StorageObject;
+import com.google.cloud.hadoop.repackaged.gcs.com.google.cloud.hadoop.gcsio.StorageResourceId;
+import io.trino.hdfs.HdfsContext;
+import io.trino.hdfs.HdfsEnvironment;
+import org.apache.hadoop.fs.Path;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+public class GcsExclusiveOutputStream
+        extends ByteArrayOutputStream
+{
+    private final Storage storage;
+    private final Path path;
+    private boolean closed;
+
+    public GcsExclusiveOutputStream(HdfsEnvironment environment, HdfsContext context, Path path)
+    {
+        this.storage = environment.createGcsStorage(context, path);
+        this.path = path;
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+
+        StorageResourceId storageResourceId = StorageResourceId.fromStringPath(path.toString());
+        Storage.Objects.Insert insert = storage.objects().insert(
+                storageResourceId.getBucketName(),
+                new StorageObject().setName(storageResourceId.getObjectName()),
+                new ByteArrayContent("application/json", toByteArray()));
```

Comment:
Compared to `GcsTransactionLogSynchronizer` we're not necessarily writing only json content through this output stream.

Reply:
Good catch. Do you know if the actual content type matters? It can't be set using the Hadoop APIs.

Reply:
https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/d570fa7a548a721bae8452c3adb6a5de6e39c714/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java#L37